tests/system/src/test/java/test/robot/javafx/scene/TextCharacterIndexTest.java

```diff
@@ -0,0 +1,456 @@
+ * 1. Create a TextFlow. Add two Text nodes, one with text and one with emojis.
+ * 2. Move the cursor from first character to last character with a random
+ *    incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testEmojiAndTextCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes, one with emojis and one with text.
+ * 2. Move the cursor from first character to last character with a random
+ *    incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testTwoEmojiNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes containing only emojis.
+ * 2. Move the cursor from first character to last character with a random
+ *    incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testWrappedTextNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes which gets wrapped.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testMultiLineTextNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes with emojis where new line is explicitly added.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testMultiLineEmojisCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes with emojis where new line is explicitly added.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ */
+
+public class TextCharacterIndexTest {
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+    static Random random;
+    static Robot robot;
+    static TextFlow textFlow;
+    static Text textOne;
+    static Text textTwo;
+    static Text emoji;
+
+    static volatile Stage stage;
+    static volatile Scene scene;
+
+    static final int WIDTH = 500;
+    static final int HEIGHT = 150;
+
+    final int Y_OFFSET = 25;
+    final int X_LEADING_OFFSET = 10;
+    final int X_TRAILING_OFFSET = 40;
+
+    boolean isLeading;
+    boolean isSurrogatePair;
+    int charIndex;
+    int insertionIndex;
+    int textFlowCharIndex;
+
+    private void mouseClick(double x, double y) {
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (scene.getWindow().getX() + scene.getX() + x),
```

Comment:
minor: local Window variable eliminates multiple invocations of getWindow()
(I am too lazy to see if javac is smart enough to optimizes it away)

Reply:
Made changes as suggested. As far as I checked, javac does not optimize this.